ads/google/a4a/utils.js

```diff
@@ -65,19 +65,85 @@ export function isGoogleAdsA4AValidEnvironment(win, element) {
-  return getAdCid(a4a).then(clientId => referrerPromise.then(referrer =>
-      buildAdUrl(
-          a4a, baseUrl, startTime, slotNumber, queryParams,
-          unboundedQueryParams, clientId, referrer)));
+  return getAdCid(a4a).then(clientId => referrerPromise.then(referrer => {
+    const slotNumber = a4a.element.getAttribute('data-amp-slot-index');
+    const win = a4a.win;
+    const documentInfo = documentInfoForDoc(a4a.element);
+    if (!win.gaGlobal) {
+      // Read by GPT for GA/GPT integration.
+      win.gaGlobal = {
+        vid: clientId,
+        hid: documentInfo.pageViewId,
+      };
+    }
+    const slotRect = a4a.getIntersectionElementLayoutBox();
+    const screen = win.screen;
+    const viewport = a4a.getViewport();
+    const viewportRect = viewport.getRect();
+    const iframeDepth = iframeNestingDepth(win);
+    const viewportSize = viewport.getSize();
+    const dtdParam = {name: 'dtd'};
+    const adElement = a4a.element;
+    const allQueryParams = queryParams.concat(
+      [
+        {
+          name: 'is_amp',
+          value: AmpAdImplementation.AMP_AD_XHR_TO_IFRAME_OR_AMP,
+        },
+        {name: 'amp_v', value: '$internalRuntimeVersion$'},
+        {name: 'd_imp', value: '1'},
+        {name: 'dt', value: startTime},
+        {name: 'ifi', value: slotNumber},
+        {name: 'adf', value: domFingerprint(adElement)},
+        {name: 'c', value: getCorrelator(win, clientId)},
+        {name: 'output', value: 'html'},
+        {name: 'nhd', value: iframeDepth},
+        {name: 'iu', value: a4a.element.getAttribute('data-ad-slot')},
+        {name: 'eid', value: a4a.element.getAttribute('data-experiment-id')},
+        {name: 'biw', value: viewportRect.width},
+        {name: 'bih', value: viewportRect.height},
+        {name: 'adx', value: slotRect.left},
+        {name: 'ady', value: slotRect.top},
+        {name: 'u_aw', value: screen ? screen.availWidth : null},
+        {name: 'u_ah', value: screen ? screen.availHeight : null},
+        {name: 'u_cd', value: screen ? screen.colorDepth : null},
+        {name: 'u_w', value: screen ? screen.width : null},
+        {name: 'u_h', value: screen ? screen.height : null},
+        {name: 'u_tz', value: -new Date().getTimezoneOffset()},
+        {name: 'u_his', value: getHistoryLength(win)},
+        {name: 'brdim', value: additionalDimensions(win, viewportSize)},
+        {name: 'isw', value: viewportSize.width},
+        {name: 'ish', value: viewportSize.height},
+        dtdParam,
+      ],
+      unboundedQueryParams,
+      [
+        {name: 'url', value: documentInfo.canonicalUrl},
+        {name: 'top', value: iframeDepth ? topWindowUrlOrDomain(win) : null},
+        {
+          name: 'loc',
+          value: win.location.href == documentInfo.canonicalUrl ?
+            null : win.location.href,
+        },
+        {name: 'ref', value: referrer},
+      ]
+    );
+    dtdParam.value = elapsedTimeWithCeiling(Date.now(), startTime);
```

Comment:
Setting dtd here means that it does not include the time required to execute buildUrl.  Instead, I suggest:
`const url = buildUrl(baseUrl, allQueryParams, MAX_URL_LENGTH - 10, {name: 'trunc', value: '1'});`
`return url += '&dtd=' + elapsedTimeWithCeiling(Date.now(), startTime);`